x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTemplateRegistryTests.java

```diff
@@ -89,7 +89,7 @@ public void tearDown() throws Exception {
     }
 
     public void testThatNonExistingComposableTemplatesAreAddedImmediately() throws Exception {
-        DiscoveryNode node = DiscoveryNodeUtils.create("node");
+        DiscoveryNode node = DiscoveryNodeUtils.builder("node").features(CONNECTOR_TEMPLATES_FEATURE).build();
```

Comment:
So, this is an issue - the feature registration is done by the plugin, which is not loaded here. This means that no historical features are specified at all, so the feature needs to be explicitly added in tests. Not great.

Having a static registration method would fix historical features, but not new feature added after FeatureService is merged - those would always need to be explicitly specified whenever `DiscoveryNode` is used directly.

Reply:
This would mean that every plugin that wants to auto-register a feature would have to go through the faff of using the java SPI stuff.

I'm not sure how much of an issue this is, given these uses are largely internal...

Reply:
> would have to go through the faff of using the java SPI stuff.

Yeah, my thoughts exactly. I'm not sure is much of an issue either, you can consider it part of mocking -- setup tests with a partial/non complete infrastructure.

Reply:
I've changed to use an SPI interface. I'm much happier with this, but we will still need a different mechanism for external plugins later

Reply:
I might be biased as I've been struggling with SPIs all week :) but I liked the non-SPI approach more. 
I feel the problem of having to declare a feature and then register it separately (potentially forgetting to register it) now is a bit worse: you have to remember to add them to a different class (and to create it in the first place).

Reply:
The issue is that the features need to be declared implicity, so that you can just use it in `DiscoveryNode` without changing anything. We also need to be able to control when features get registered, and potentially re-register them for tests etc. Hence the use of an interface implementation.

I think one class per plugin that is responsible for all feature registrations isn't too bad - it only needs to be set up once.

Reply:
> I think one class per plugin that is responsible for all feature registrations isn't too bad - it only needs to be set up once.

That's true, we might be worrying for no reason